contracts/vested-escrow/VeERC20v2.sol

```diff
@@ -0,0 +1,208 @@
+
+/*
+ * @title The contact enables the storage of erc20 tokens under the linear time-vesting with the cliff time-lock.
+ * @notice During cliff period all tokens are locked so the user can't claim it.
+ * After the cliff period is over token gets gradually unlock every second until the vesting is over.
+ *
+ * For example, if cliff is 3 months and vesting is 12 months,
+ *      during the first 3 mouths user can't claim anything
+ *      after 4 months (i.e. 1 month after the cliff) user can claim 1/9 of all tokens
+ *      after 6 months user can claim 3/9 of all tokens etc.
+ * It's possible to claim tokens by parts during vesting, or claim all at once when vesting if over.
+ *
+ * To make balance visible in the erc20 wallets, the contact "looks like" erc20 token by implementing its interface
+ * however all non-view methods such as transfer or approve aren't active and will be reverted.
+*/
+contract VeERC20v2 is SuVoteToken, ERC20BurnableUpgradeable, IveERC20v2 {
+    using SafeERC20Upgradeable for ERC20Upgradeable;
+
+    ERC20Upgradeable public LOCKED_TOKEN;
+    uint32 public tgeTimestamp;
+
+    // Amount of seconds while tokens would be completely locked.
+    uint32 public cliffSeconds;
+
+    // Amount of seconds when vesting would be over. Starts from cliff.
+    uint32 public vestingSeconds;
+
+    // ratio/1e18 ⊂ [0..1] that indicates how many tokens are going to be unlocked during TGE
+    // uint64 is enough because log2(1e18) ~= 60
+    uint64 public tgeUnlockRatio1e18;
+
+    // how frequently token are going to be unlocked after the cliff.
+    uint32 public vestingFrequencySeconds;
+
+    mapping(address => uint256) public alreadyWithdrawn;
+
+    function initialize(
+        address _accessControlSingleton,
+        ERC20Upgradeable _lockedToken,
+        uint32 _maxTgeTimestamp
+    ) initializer public {
+        string memory veSymbol = string.concat("ve", _lockedToken.symbol());
+        __SuVoteToken__init(_accessControlSingleton, veSymbol);
+        __ERC20_init(string.concat("vested escrow ", _lockedToken.name()), veSymbol);
+
+        LOCKED_TOKEN = _lockedToken;
+        tgeTimestamp = _maxTgeTimestamp;
+
+        cliffSeconds = 6 * 30 days;
+        vestingSeconds = 24 * 30 days;
+        vestingFrequencySeconds = 6 * 30 days;
+        tgeUnlockRatio1e18 = 10 * 1e16; // 10%
+    }
+
+    function updateTimestamps(
+        uint32 newTgeTimestamp,
+        uint32 newCliffSeconds,
+        uint32 newVestingSeconds,
+        uint64 newTgeUnlockRatio,
+        uint32 newVestingFrequencySeconds
+    ) external onlyRole(ADMIN_ROLE) {
+        if (newTgeTimestamp < uint32(block.timestamp)) revert TGEInPastError();
+        if (newTgeUnlockRatio > 1e18) revert BadUnlockRatio();
+        if (newVestingSeconds == 0) revert BadVestingTimestamps();
+        if (
+            newVestingSeconds == 0 ||
+            newVestingSeconds / newVestingFrequencySeconds * newVestingFrequencySeconds != newVestingSeconds
```

Comment:
not a cocern, just curious what's the advantage of this in comparison with newVestingSeconds % newVestingFrequencySeconds == 0?
also not sure it needed, but I still have nightname flashback after years of C++ so use brackets everywhere